RecoBTag/DeepFlavour/plugins/DeepFlavourJetTagProducer.cc

```diff
@@ -0,0 +1,234 @@
+    typedef reco::JetTagCollection JetTagCollection;
+
+    virtual void beginStream(edm::StreamID) override {}
+    virtual void produce(edm::Event&, const edm::EventSetup&) override;
+    virtual void endStream() override {}
+
+    const edm::EDGetTokenT< TagInfoCollection > src_;
+    edm::FileInPath graph_path_;
+    std::vector<std::pair<std::string,std::vector<unsigned int>>> flav_pairs_;
+    std::vector<std::string> input_names_;
+    std::vector<std::string> output_names_;
+
+    // graph for TF evaluation
+    tf::Graph graph_;
+    // vector of tensors for inputs, outputs and scalar learning phase 
+    std::vector<tf::Tensor*> dnn_inputs_;
+    std::vector<tf::Tensor*> dnn_outputs_;
+    std::vector<tf::Tensor*> dnn_lp_tensors_;
+};
+
+DeepFlavourJetTagProducer::DeepFlavourJetTagProducer(const edm::ParameterSet& iConfig) :
+  src_(consumes<TagInfoCollection>(iConfig.getParameter<edm::InputTag>("src"))),
+  graph_path_(iConfig.getParameter<edm::FileInPath>("graph_path")),
+  input_names_(iConfig.getParameter<std::vector<std::string>>("input_names")),
+  output_names_(iConfig.getParameter<std::vector<std::string>>("output_names")),
+  graph_(graph_path_.fullPath().substr(0, graph_path_.fullPath().find_last_of("/")), "serve")
+{
+
+  // get output names from flav_table
+  const auto & flav_pset = iConfig.getParameter<edm::ParameterSet>("flav_table");
+  for (const auto flav_pair : flav_pset.tbl()) {
+    const auto & flav_name = flav_pair.first;
+    flav_pairs_.emplace_back(flav_name,
+                             flav_pset.getParameter<std::vector<unsigned int>>(flav_name));
+  }
+
+  for (const auto & flav_pair : flav_pairs_) {
+    produces<JetTagCollection>(flav_pair.first);
+  }
+
+  for (std::size_t i=0; i<input_names_.size(); i++) {
+    dnn_inputs_.push_back(new tf::Tensor());
+  }
+
+  // required because of batch norm
+  // names for the learing phase placeholders (to init and set as false)
+  const auto & lp_names = iConfig.getParameter<std::vector<std::string>>("lp_names");
+  for (const auto & lp_name : lp_names) {
+    dnn_lp_tensors_.push_back(new tf::Tensor(0, 0, TF_BOOL));
+    *(dnn_lp_tensors_.back()->getPtr<bool>()) = false;
+    graph_.defineInput(dnn_lp_tensors_.back(), lp_name);
+  }
+
+  for (const auto & output_name : output_names_) {
+    dnn_outputs_.push_back(new tf::Tensor());
+    graph_.defineOutput(dnn_outputs_.back(), output_name);
+  }
+
+}
+
+
+DeepFlavourJetTagProducer::~DeepFlavourJetTagProducer()
+{
+  // cleanup tensor vectors
+  while (!dnn_inputs_.empty()) {
+    delete dnn_inputs_.back();
+    dnn_inputs_.pop_back();
+  }
+  while (!dnn_outputs_.empty()) {
+    delete dnn_outputs_.back();
+    dnn_outputs_.pop_back();
+  }
+  while (!dnn_lp_tensors_.empty()) {
+    delete dnn_lp_tensors_.back();
+    dnn_lp_tensors_.pop_back();
+  }
+}
+
+void DeepFlavourJetTagProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions)
+{
+}
```

Comment:
better fill with something appropriate, with default parameters

Reply:
will be filled